File: VoiceAIApp/app/src/main/java/com/ubudy/voiceai/TokenService.kt
package com.ubudy.voiceai

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query

data class TokenResponse(val token: String, val url: String)

data class UserInfo(
    val name: String = "",
    val subject: String = "",
    val grade: String = "",
    val language: String = "English"
)

interface TokenService {
    @GET("/token")
    suspend fun getToken(
        @Query("name") name: String = "",
        @Query("subject") subject: String = "",
        @Query("grade") grade: String = "",
        @Query("language") language: String = "English"
    ): TokenResponse
}

object TokenClient {
    private const val BASE_URL = "https://advancedvoiceagent.xappy.io"

    val service: TokenService by lazy {
        Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(TokenService::class.java)
    }
}

File: VoiceAIApp-iOS/VoiceAIApp/TokenService.swift
import Foundation

struct TokenResponse: Decodable {
    let token: String
    let url: String
}

struct UserInfo {
    var name: String = ""
    var subject: String = ""
    var grade: String = ""
    var language: String = "English"
    var type: String = ""
}

enum TokenService {
    static let baseURL = "https://advancedvoiceagent.xappy.io"

    static func fetchToken(userInfo: UserInfo) async throws -> TokenResponse {
        var components = URLComponents(string: "\(baseURL)/token")!
        components.queryItems = [
            URLQueryItem(name: "name", value: userInfo.name),
            URLQueryItem(name: "subject", value: userInfo.subject),
            URLQueryItem(name: "grade", value: userInfo.grade),
            URLQueryItem(name: "language", value: userInfo.language),
            URLQueryItem(name: "type", value: userInfo.type),
        ]
        guard let url = components.url else {
            throw URLError(.badURL)
        }
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode(TokenResponse.self, from: data)
    }
}
